File: packages/alteration/src/main/java/fdit/alteration/api/AlterationAPI.java
package fdit.alteration.api;

import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import fdit.alteration.core.engine.EngineManager;
import fdit.alteration.core.engine.EngineParameters;
import fdit.alteration.core.incident.IncidentDeserializer;
import fdit.alteration.core.incident.Recording;
import fdit.alteration.core.incident.Sensor;
import fdit.alteration.core.logging.ActionLogger;
import org.apache.commons.cli.CommandLine;

import java.io.File;
import java.io.FileNotFoundException;

import static com.google.common.io.Files.copy;
import static org.apache.commons.io.FilenameUtils.getExtension;
import static org.apache.commons.io.FilenameUtils.separatorsToSystem;

public class AlterationAPI {

    private AlterationAPI() {

    }

    public static void startAlteration(final File destination, final File incidentFile) throws Exception {
        startAlteration(destination, incidentFile, false, new EngineParameters(), "modified", "");
    }

    public static void startAlteration(final File destination,
                                       final File incidentFile,
                                       final boolean labeled) throws Exception {
        startAlteration(destination, incidentFile, false, labeled);
    }

    public static void startAlteration(final File destination,
                                       final File incidentFile,
                                       final boolean logResults,
                                       final boolean labeled) throws Exception {
        startAlteration(destination, incidentFile, logResults, new EngineParameters(labeled), "modified", "");
    }

    public static void startAlteration(final File destination,
                                       final File incidentFile,
                                       final boolean logResults,
                                       final boolean labeled,
                                       final String prefix,
                                       final String suffix) throws Exception {
        startAlteration(destination, incidentFile, logResults, new EngineParameters(labeled), prefix, suffix);
    }


    public static void startAlteration(final File destination,
                                       final File incidentFile,
                                       final boolean logResults,
                                       final CommandLine commandLine,
                                       final String prefix,
                                       final String suffix) throws Exception {
        startAlteration(destination, incidentFile, logResults, new EngineParameters(commandLine), prefix, suffix);
    }

    public static void startAlteration(final File destination,
                                       final File incidentFile,
                                       final boolean logResults,
                                       final EngineParameters parameters,
                                       final String prefix,
                                       final String suffix) throws Exception {
        for (final Sensor sensor : new IncidentDeserializer(incidentFile).deserialize(new XmlMapper()).getSensors()) {
            final File recordingFile = new File(incidentFile.getParent() +
                    separatorsToSystem("/") +
                    sensor.getRecord());
            if (!recordingFile.exists() || !recordingFile.isFile()) {
                throw new FileNotFoundException(recordingFile.getAbsolutePath());
            }
            final Recording recording = new Recording(recordingFile, sensor.getFirstDate());
            final ActionLogger logger = new ActionLogger();
            final EngineManager engineManager = new EngineManager(recording, sensor.getActions(), logger, parameters);
            recording.setFile(engineManager.run());
            copy(recording.getFile(), new File(
                    destination,
                    prefix + "_" +
                            sensor.getsID() + "_" +
                            suffix + "."
                            + getExtension(recordingFile.getName())));
            if (logResults) {
                logger.print(new File(recordingFile.getParent(), "metrics.csv"));
            }
        }
    }
}

File: fdit-ts/src/fdit/alteration/api/AlterationAPI.ts
import * as fs from 'fs'
import { copyFileSync } from 'fs'

import { EngineManager } from '../core/engine/EngineManager'
import { EngineParameters } from '../core/engine/EngineParameters'
import { Recording } from '../core/incident/Recording'
import { Sensor } from '../core/incident/Sensor'
import { ActionLogger } from '../core/logging/ActionLogger'

export class AlterationAPI {
  private constructor() {}

  public static startAlteration(
    destination: string,
    incidentFilePath: string,
    options?: {
      labeled?: boolean
      logResults?: boolean
      commandLine?: string
      prefix?: string
      suffix?: string
    }
  ): void {
    console.log('startAlteration')
    const {
      labeled = false,
      logResults = false,
      commandLine = '',
      prefix = 'modified',
      suffix = '',
    } = options || {}

    const incidentData = fs.readFileSync(incidentFilePath, 'utf-8')
    const incident: Incident = JSON.parse(incidentData)
    const sensors: Sensor[] = incident.getSensors()

    let argCount = 2 // par défaut

    if (labeled && logResults && commandLine && prefix && suffix) {
      argCount = 6 // toutes les options
      console.log('startAlteration6args')
    } else {
      console.log('startAlteration2args')
    }

    for (const sensor of sensors) {
      const recordingFilePath = `${incidentFilePath}/${sensor.getRecord()}`
      if (
        !fs.existsSync(recordingFilePath) ||
        !fs.statSync(recordingFilePath).isFile()
      ) {
        console.log('FileNotFound')
        throw new Error(`File not found: ${recordingFilePath}`)
      }

      const recording = new Recording(recordingFilePath, sensor.getFirstDate())
      const logger = new ActionLogger()
      console.log('Nombre action : ' + sensor.getActions().length)

      const engineManager = new EngineManager(
        recording,
        sensor.getActions(),
        logger,
        new EngineParameters(labeled, commandLine)
      )

      recording.setFile(engineManager.run())

      const newFileName =
        prefix +
        '_' +
        sensor.getsID() +
        '_' +
        suffix +
        '.' +
        this.getFileExtension(recordingFilePath)

      copyFileSync(recording.getFile(), `${destination}/${newFileName}`)

      if (logResults) {
        logger.print(`${recordingFilePath}/metrics.csv`)
      }

      console.log('in for')
    }

    console.log('For finish')
  }

  private static getFileExtension(name: string): string {
    const blocks = name.split('.')
    if (blocks.length > 1) {
      return blocks[blocks.length - 1]
    }
    return ''
  }
}
